File: ros/src/system/autoware_health_checker/src/node_status_publisher.cpp
/*
 * Copyright 2019 Autoware Foundation. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *
 * v1.0 Masaya Kataoka
 */

#include <autoware_health_checker/node_status_publisher.h>

namespace autoware_health_checker {
NodeStatusPublisher::NodeStatusPublisher(ros::NodeHandle nh,
                                         ros::NodeHandle pnh) {
  node_activated_ = false;
  ros_ok_ = true;
  nh_ = nh;
  pnh_ = pnh;
  status_pub_ =
      nh_.advertise<autoware_system_msgs::NodeStatus>("node_status", 10);
}

NodeStatusPublisher::~NodeStatusPublisher() {
  ros_ok_ = false;
  publish_thread_.join();
}

void NodeStatusPublisher::publishStatus() {
  ros::Rate rate = ros::Rate(autoware_health_checker::UPDATE_RATE);
  while (ros_ok_) {
    mtx_.lock();
    autoware_system_msgs::NodeStatus status;
    status.node_activated = node_activated_;
    ros::Time now = ros::Time::now();
    status.header.stamp = now;
    status.node_name = ros::this_node::getName();
    std::vector<std::string> checker_keys = getRateCheckerKeys();
    // iterate Rate checker and publish rate_check result
    for (auto key_itr = checker_keys.begin(); key_itr != checker_keys.end();
         key_itr++) {
      autoware_system_msgs::DiagnosticStatusArray diag_array;
      autoware_system_msgs::DiagnosticStatus diag;
      diag.type = autoware_system_msgs::DiagnosticStatus::RATE_IS_SLOW;
      std::pair<uint8_t, double> result =
          rate_checkers_[*key_itr]->getErrorLevelAndRate();
      diag.level = result.first;
      diag.key = *key_itr;
      diag.value = doubeToJson(result.second);
      diag.description = rate_checkers_[*key_itr]->description;
      diag.header.stamp = now;
      diag_array.status.push_back(diag);
      status.status.push_back(diag_array);
    }
    // iterate Diagnostic Buffer and publish all diagnostic data
    std::vector<std::string> keys = getKeys();
    for (auto key_itr = keys.begin(); key_itr != keys.end(); key_itr++) {
      status.status.push_back(diag_buffers_[*key_itr]->getAndClearData());
    }
    status_pub_.publish(status);
    mtx_.unlock();
    rate.sleep();
  }
  return;
}

void NodeStatusPublisher::ENABLE() {
  publish_thread_ = boost::thread(boost::bind(&NodeStatusPublisher::publishStatus, this));
  return;
}

std::vector<std::string> NodeStatusPublisher::getKeys() {
  std::vector<std::string> keys;
  std::vector<std::string> checker_keys = getRateCheckerKeys();
  std::pair<std::string, std::shared_ptr<DiagBuffer>> buf_itr;
  BOOST_FOREACH (buf_itr, diag_buffers_) {
    bool matched = false;
    for (auto checker_key_itr = checker_keys.begin();
         checker_key_itr != checker_keys.end(); checker_key_itr++) {
      if (*checker_key_itr == buf_itr.first) {
        matched = true;
      }
    }
    if (!matched) {
      keys.push_back(buf_itr.first);
    }
  }
  return keys;
}

std::vector<std::string> NodeStatusPublisher::getRateCheckerKeys() {
  std::vector<std::string> keys;
  std::pair<std::string, std::shared_ptr<RateChecker>> checker_itr;
  BOOST_FOREACH (checker_itr, rate_checkers_) {
    keys.push_back(checker_itr.first);
  }
  return keys;
}

bool NodeStatusPublisher::keyExist(std::string key) {
  if (diag_buffers_.count(key) == 0) {
    return false;
  }
  return true;
}

// add New Diagnostic Buffer if the key does not exist
void NodeStatusPublisher::addNewBuffer(std::string key, uint8_t type,
                                       std::string description) {
  if (!keyExist(key)) {
    std::shared_ptr<DiagBuffer> buf_ptr = std::make_shared<DiagBuffer>(
        key, type, description, autoware_health_checker::BUFFER_LENGTH);
    diag_buffers_[key] = buf_ptr;
  }
  return;
}

uint8_t NodeStatusPublisher::CHECK_MIN_VALUE(std::string key, double value,
                                             double warn_value,
                                             double error_value,
                                             double fatal_value,
                                             std::string description) {
  addNewBuffer(key, autoware_system_msgs::DiagnosticStatus::OUT_OF_RANGE,
               description);
  autoware_system_msgs::DiagnosticStatus new_status;
  new_status.type = autoware_system_msgs::DiagnosticStatus::OUT_OF_RANGE;
  if (value < fatal_value) {
    new_status.level = autoware_system_msgs::DiagnosticStatus::FATAL;
  } else if (value < error_value) {
    new_status.level = autoware_system_msgs::DiagnosticStatus::ERROR;
  } else if (value < warn_value) {
    new_status.level = autoware_system_msgs::DiagnosticStatus::WARN;
  } else {
    new_status.level = autoware_system_msgs::DiagnosticStatus::OK;
  }
  new_status.description = description;
  new_status.value = doubeToJson(value);
  new_status.header.stamp = ros::Time::now();
  diag_buffers_[key]->addDiag(new_status);
  return new_status.level;
}

uint8_t NodeStatusPublisher::CHECK_MAX_VALUE(std::string key, double value,
                                             double warn_value,
                                             double error_value,
                                             double fatal_value,
                                             std::string description) {
  addNewBuffer(key, autoware_system_msgs::DiagnosticStatus::OUT_OF_RANGE,
               description);
  autoware_system_msgs::DiagnosticStatus new_status;
  new_status.type = autoware_system_msgs::DiagnosticStatus::OUT_OF_RANGE;
  if (value > fatal_value) {
    new_status.level = autoware_system_msgs::DiagnosticStatus::FATAL;
  } else if (value > error_value) {
    new_status.level = autoware_system_msgs::DiagnosticStatus::ERROR;
  } else if (value > warn_value) {
    new_status.level = autoware_system_msgs::DiagnosticStatus::WARN;
  } else {
    new_status.level = autoware_system_msgs::DiagnosticStatus::OK;
  }
  new_status.description = description;
  new_status.value = doubeToJson(value);
  new_status.header.stamp = ros::Time::now();
  diag_buffers_[key]->addDiag(new_status);
  return new_status.level;
}

uint8_t NodeStatusPublisher::CHECK_RANGE(std::string key, double value,
                                         std::pair<double, double> warn_value,
                                         std::pair<double, double> error_value,
                                         std::pair<double, double> fatal_value,
                                         std::string description) {
  addNewBuffer(key, autoware_system_msgs::DiagnosticStatus::OUT_OF_RANGE,
               description);
  autoware_system_msgs::DiagnosticStatus new_status;
  new_status.type = autoware_system_msgs::DiagnosticStatus::OUT_OF_RANGE;
  if (value < fatal_value.first || value > fatal_value.second) {
    new_status.level = autoware_system_msgs::DiagnosticStatus::FATAL;
  } else if (value < error_value.first || value > error_value.second) {
    new_status.level = autoware_system_msgs::DiagnosticStatus::ERROR;
  } else if (value < warn_value.first || value > warn_value.second) {
    new_status.level = autoware_system_msgs::DiagnosticStatus::WARN;
  } else {
    new_status.level = autoware_system_msgs::DiagnosticStatus::OK;
  }
  new_status.value = doubeToJson(value);
  new_status.description = description;
  new_status.header.stamp = ros::Time::now();
  diag_buffers_[key]->addDiag(new_status);
  return new_status.level;
}

void NodeStatusPublisher::CHECK_RATE(std::string key, double warn_rate,
                                     double error_rate, double fatal_rate,
                                     std::string description) {
  if (!keyExist(key)) {
    std::shared_ptr<RateChecker> checker_ptr = std::make_shared<RateChecker>(
        autoware_health_checker::BUFFER_LENGTH, warn_rate, error_rate,
        fatal_rate, description);
    rate_checkers_[key] = checker_ptr;
  }
  addNewBuffer(key, autoware_system_msgs::DiagnosticStatus::RATE_IS_SLOW,
               description);
  rate_checkers_[key]->check();
  return;
}

std::string NodeStatusPublisher::doubeToJson(double value) {
  using namespace boost::property_tree;
  std::stringstream ss;
  ptree pt;
  pt.put("value.double", value);
  write_json(ss, pt);
  return ss.str();
}
}

File: ros/src/system/autoware_health_checker/scripts/autoware_health_checker/node_status_publisher.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Copyright 2015-2019 Autoware Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import rospy
from autoware_system_msgs.msg import NodeStatus
from constants import autoware_health_checker
from rate_checker import RateChecker
from autoware_system_msgs.msg import DiagnosticStatusArray
from autoware_system_msgs.msg import DiagnosticStatus
from diag_buffer import DiagBuffer
import json
import threading
from threading import Lock

class NodeStatusPublisher:
    def __init__(self):
        self.__node_activated = False
        self.__status_pub = rospy.Publisher('node_status', NodeStatus, queue_size=10)
        self.__lock = Lock()
        self.__rate_checkers = {}
        self.__diag_buffers = {}

    def publishStatus(self):
        rate = rospy.Rate(autoware_health_checker.UPDATE_RATE)
        while not rospy.is_shutdown():
            with self.__lock:
                status = NodeStatus()
                status.node_activated = self.__node_activated
                now = rospy.Time.now()
                status.header.stamp = now
                status.node_name = rospy.get_name()
                checker_keys = self.getRateCheckerKeys()
                #  iterate Rate checker and publish rate_check result
                for key_itr in checker_keys:
                    diag_array = DiagnosticStatusArray()
                    diag       = DiagnosticStatus()
                    diag.type = DiagnosticStatus.RATE_IS_SLOW
                    result = self.__rate_checkers[key_itr].getErrorLevelAndRate()
                    diag.level = result[0]
                    diag.key = key_itr
                    diag.value = self.doubleToJson(result[1])
                    diag.description = self.__rate_checkers[key_itr].description
                    diag.header.stamp = now
                    diag_array.status.append(diag)
                    status.status.append(diag_array)
                keys = self.getKeys()
                for key_itr in keys:
                    status.status.append(self.__diag_buffers[key_itr].getAndClearData())
                self.__status_pub.publish(status)
            rate.sleep()
        return

    def ENABLE(self):
        publish_thread = threading.Thread(target=self.publishStatus)
        publish_thread.start()

    def getKeys(self):
        keys = []
        checker_keys = self.getRateCheckerKeys()
        for buf_itr_first,buf_itr_second in self.__diag_buffers.items():
            matched = False
            for checker_key_itr in checker_keys:
                if checker_key_itr == buf_itr_first:
                    matched = True
            if not matched:
                keys.append(buf_itr_first)
        return keys

    def getRateCheckerKeys(self):
        keys = []
        for checker_itr_first, checker_itr_second in self.__rate_checkers.items():
            keys.append(checker_itr_first)
        return keys

    def keyExist(self, key):
        return key in self.__diag_buffers

    def addNewBuffer(self, key, type, description):
        if not self.keyExist(key):
            buf = DiagBuffer(key, type, description, autoware_health_checker.BUFFER_LENGTH)
            self.__diag_buffers[key] = buf

    def CHECK_MIN_VALUE(self, key, value, warn_value, error_value, fatal_value, description):
        self.addNewBuffer(key, DiagnosticStatus.OUT_OF_RANGE, description)
        new_status = DiagnosticStatus()
        new_status.type = DiagnosticStatus.OUT_OF_RANGE
        if value < fatal_value:
            new_status.level = DiagnosticStatus.FATAL
        elif value < error_value:
            new_status.level = DiagnosticStatus.ERROR
        elif value < warn_value:
            new_status.level = DiagnosticStatus.WARN
        else:
            new_status.level = DiagnosticStatus.OK
        new_status.description = description
        new_status.value = self.doubleToJson(value)
        new_status.header.stamp = rospy.Time.now()
        self.__diag_buffers[key].addDiag(new_status)
        return new_status.level

    def CHECK_MAX_VALUE(self, key, value, warn_value, error_value, fatal_value, description):
        self.addNewBuffer(key, DiagnosticStatus.OUT_OF_RANGE, description)
        new_status = DiagnosticStatus()
        new_status.type = DiagnosticStatus.OUT_OF_RANGE
        if value > fatal_value:
            new_status.level = DiagnosticStatus.FATAL
        elif value > error_value:
            new_status.level = DiagnosticStatus.ERROR
        elif value > warn_value:
            new_status.level = DiagnosticStatus.WARN
        else:
            new_status.level = DiagnosticStatus.OK
        new_status.description = description
        new_status.value = self.doubleToJson(value)
        new_status.header.stamp = rospy.Time.now()
        self.__diag_buffers[key].addDiag(new_status)
        return new_status.level

    def CHECK_RANGE(self,key, value, warn_value, error_value, fatal_value, description):
        self.addNewBuffer(key, DiagnosticStatus.OUT_OF_RANGE, description)
        new_status = DiagnosticStatus()
        new_status.type = DiagnosticStatus.OUT_OF_RANGE
        if value < fatal_value[0] or value > fatal_value[1]:
            new_status.level = DiagnosticStatus.FATAL
        elif value < error_value[0] or value > error_value[1]:
            new_status.level = DiagnosticStatus.ERROR
        elif value < warn_value[0] or value > warn_value[1]:
            new_status.level = DiagnosticStatus.WARN
        else:
            new_status.level = DiagnosticStatus.OK
        new_status.description = description
        new_status.value = self.doubleToJson(value)
        new_status.header.stamp = rospy.Time.now()
        self.__diag_buffers[key].addDiag(new_status)
        return new_status.level

    def CHECK_VALUE(self,key, value, check_func, value_json_func, description):
        self.addNewBuffer(key, DiagnosticStatus.OUT_OF_RANGE, description)
        check_result = check_func(value)
        pt = value_json_func(value)
        ss = json.dumps(pt)
        new_status = DiagnosticStatus()
        new_status.type = DiagnosticStatus.OUT_OF_RANGE
        new_status.level= check_result
        new_status.description = description
        new_status.description = ss
        new_status.header.stamp = rospy.Time.now()
        self.__diag_buffers[key].addDiag(new_status)
        return new_status.level


    def CHECK_RATE(self,key, warn_rate, error_rate, fatal_rate, description):
        if not self.keyExist(key):
            self.__rate_checkers[key] = RateChecker(autoware_health_checker.BUFFER_LENGTH,warn_rate,error_rate,fatal_rate,description)
        self.addNewBuffer(key,DiagnosticStatus.RATE_IS_SLOW,description)
        self.__rate_checkers[key].check()

    def NODE_ACTIVATE(self):
        with self.__lock:
            self.__node_activated = True

    def NODE_DEACTIVATE(self):
        with self.__lock:
            self.__node_activated  = False

    def getNodeStatus(self):
        return self.__node_activated

    def doubleToJson(self,value):
        return json.dumps({"value.double": value})
